x/microtx/keeper/msg_server.go

```diff
@@ -62,49 +62,87 @@ func (m msgServer) Microtx(c context.Context, msg *types.MsgMicrotx) (*types.Msg
+	// nolint: exhaustruct
+	if !k.gasfreeKeeper.IsGasFreeMsgType(ctx, sdk.MsgTypeURL(&types.MsgMicrotx{})) {
+		collected, err := k.DeductMicrotxFee(ctx, sender, amount)
+		if err != nil {
+			return sdkerrors.Wrap(err, "unable to collect MsgMicrotx fees")
+		}
+		ctx.EventManager().EmitEvent(types.NewEventMicrotxFeeCollected(sender.String(), *collected))
 	}
 
+	// Perform the transfer now that fees have been collected
 	err = k.bankKeeper.SendCoins(ctx, sender, receiver, sdk.NewCoins(amount))
-
 	if err != nil {
 		return sdkerrors.Wrap(err, "unable to send tokens via the bank module")
 	}
 
 	// Emit an event for the block's event log
 	ctx.EventManager().EmitEvent(
-		types.NewEventMicrotx(sender.String(), receiver.String(), amount, *feeCollected),
+		types.NewEventMicrotx(sender.String(), receiver.String(), amount),
 	)
 
 	// Detect if this is a Liquid Infrastructure Account and if so
 	// migrate balances to the NFT if the amounts are in excess of any configured threshold
 	k.Logger(ctx).Debug("Detecting and funneling excess balances for liquid infrastructure accounts")
-	if err := k.RedirectLiquidAccountExcessBalance(ctx, receiver, erc20Amount); err != nil {
+	if err := k.RedirectLiquidAccountExcessBalance(ctx, receiver, erc20Address); err != nil {
 		return sdkerrors.Wrapf(err, "failed to redirect excess balance")
 	}
 
 	return nil
 }
 
-// checkAndDeductSendToEthFees asserts that the minimum chainFee has been met for the given sendAmount
+func (k Keeper) ValidateAndGetERC20Address(ctx sdk.Context, amount sdk.Coin) (common.Address, error) {
+	var erc20Address common.Address
+	// The native token is automatically usable within the EVM
+	if amount.Denom != config.BaseDenom {
+		// Ensure the input tokens are actively registered as an ERC20-convertible token
+		pair, found := k.erc20Keeper.GetTokenPair(ctx, k.erc20Keeper.GetTokenPairID(ctx, amount.Denom))
+		if !found {
+			return common.Address{}, sdkerrors.Wrapf(types.ErrInvalidMicrotx, "token %v is not registered as an erc20, only evm-compatible tokens may be used", amount.Denom)
+		}
+		if !pair.Enabled {
+			return common.Address{}, sdkerrors.Wrapf(types.ErrInvalidMicrotx, "token %v is registered as an erc20 (%v), but the pair is not enabled", amount.Denom, pair.Erc20Address)
+		}
+		// Collect the ERC20 address for later use in funneling
+		erc20Address = common.HexToAddress(pair.Erc20Address)
+	}
+
+	return erc20Address, nil
+}
+
+// DeductMsgMicrotxFee is expected to be called from the AnteHandler to deduct the fee for the Msg
+// It is possible for MsgMicrotx to not be a gasfree message type, since governance controls the list,
+// in that case the fee should be deducted in the Msg handler
+//
+// WARNING: Do **NOT** call this from the MsgMicrotx handler, as it will result in bad event logs, call DeductMicrotxFee instead
+func (k Keeper) DeductMsgMicrotxFee(ctx sdk.Context, msg *types.MsgMicrotx) (feeCollected *sdk.Coin, err error) {
```

Comment:
could this be made private?

Reply:
No, we have to call it from the ChargeGasfreeFeesDecorator so we need some means to charge those fees in both the Msg handler and in the AnteDecorator.